### ISSUE TITLE
How to assign function s_repeat's parameter 'variable'?

### ISSUE THREAD

**Issue by JsHuang** [jtpereyda/boofuzz#333]
I want the repeater's repeat number is the value of a specific primitve variable, and I found the `s_repeat` has a `variable` parameter which in the Doc decribed as below:
> variable (Sulley Integer Primitive) – (Optional, def=None) An integer primitive
which will specify the number of repitions

But how to assign this parameter, any examples? 
Thanks.

**Comment by SR4ven** [comment]
I did some testing and the `variable` argument of `s_repeate` appears to do the opposite of what you want.
As it says in the help text, the current value of an integer primitive will be used to set the number of repetitions. It will not overwrite the integer primitive with the current repetition number.
Your goal can be archived by using callbacks:

```python
def update_reps(target, fuzz_data_logger, node, session, edge, *args, **kwargs):
    node.names['reps']._value = node.names['repeater'].current_reps
    return node.render()


def reps():
    s_initialize("HTTP COOKIE")
    s_static("GET / HTTP/1.1\r\n")

    with s_block("cookie"):
        s_static("Cookie: ")
        s_string("auth", fuzzable=False)
        s_delim("=", fuzzable=False)
        s_dword(name="reps", value=1, output_format='ascii', fuzzable=False)
        s_static("\r\n")

    s_repeat(name="repeater", block_name="cookie", max_reps=10, step=1)
    s_static("\r\n")

    session = Session(
        target=Target(connection=SocketConnection("127.0.0.1", 80, proto='tcp')))

    sess.connect(s_get("HTTP COOKIE"), callback=update_reps)
    sess.fuzz()
```

If the primitive which is being modified is inside the repeated block, all but the latest repetition will lag behind one iteration.
```http
GET / HTTP/1.1\r\nCookie: auth=3\r\nCookie: auth=2\r\nCookie: auth=2\r\nCookie: auth=2\r\n\r\n
```

During the testing I found an issue with the `variable` argument of the `s_repeat` block. Currently it's not resolving block names, which makes using it kind of ugly...
This doesn't work:
```python
s_dword(name="reps", value=1, output_format='ascii', fuzzable=False)
s_repeat(name="repeater", block_name="cookie", variable="reps")
```
and raises `Attempt to bind the repeater for block reps to a non-integer primitive!`.
I will open a separate issue about this.

**Comment by JsHuang** [comment]
Thanks, it's better to update the help doc about the description of  `variable` parameter.

**Comment by jtpereyda** [comment]
@JsHuang any suggestions for how to reword that description?

**Comment by JsHuang** [comment]
@

> @JsHuang any suggestions for how to reword that description?

I really don't know how to use this parameter, may be add an usage example ?

**Comment by SR4ven** [comment]
If understand the usage of the primitive correctly, there's nothing wrong with the current description.

I'd expect the following example to repeat the `body` block as many times as the current value of `int_value` represents.
```python
s_int(
    name='int_value',
    value=1,
    signed=False,
    fuzzable=True,
)
if s_block_start("body"):
    s_delim(" ")
    s_static("\r\n\r\n")
s_block_end()
s_repeat(name="repeater", block_name="body", variable="int_value")
```
Instead, the following error is risen:
`boofuzz.exception.SullyRuntimeError: Attempt to bind the repeater for block body to a non-integer primitive!`

Or is this the correct way to use this feature?
```python
s_repeat(
    name="repeater",
    block_name="body",
    variable=s_int(
        value=1, signed=False, fuzzable=True, name="int_value"
    ),
    max_reps=10,
    step=1,
)
```
Looks a little weird to me but it works fine.

**Comment by jtpereyda** [comment]
Hmm, I believe you need to get it a reference to the object. It would be something like:
```python
s_repeat(
    name="repeater",
    block_name="body",
    variable=REQUESTS["requestname"].names["int_value"],
    max_reps=10,
    step=1,
)
```
Something like that.

**Comment by SR4ven** [comment]
> Something like that.

Nope, that results in a type error at https://github.com/jtpereyda/boofuzz/blob/74dd08c783c2301081751fe793c94d8b36c62730/boofuzz/blocks/repeat.py#L157
because `self.variable.render()` return a bytes string like `\x00\x00\x00\x00`. The problem is, that we cannot convert this to an int easily (not that I know at least) because we don't know size and padding.

One way to get the current value as int would be to directly access `self.variable._value`, which results in the expected behaviour.
But running the session quickly results in a memory error. My guess is that the int values are getting big quickly, resulting in huge string sizes. This raises another question: What is the use of this option if it's more or less unusable?

On top my example here
```python
s_repeat(
    name="repeater",
    block_name="body",
    variable=s_int(
        value=1, signed=False, fuzzable=True, name="int_value"
    ),
    max_reps=10,
    step=1,
)
```
results in the int primitive being appended to the assigned block. But the repetitions of the block are not affected by it in any way, even though it is being mutated.

I have to say, this option raises more and more questions...

**Comment by jtpereyda** [comment]
@SR4ven You're right about using `_value`, that would be the right way to go even though the variable is marked as private. We could update IFuzzable to have a non-rendered value property, too.

I think if you use `_value`, and use the assignment style I previously posted (`variable=REQUESTS["requestname"].names["int_value"],`), it should work. Maybe.

Disclaimer: I've never used this one in any practical way myself.
